framework/src/meshmodifiers/LowerDBlockFromSideset.C

```diff
@@ -0,0 +1,102 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "LowerDBlockFromSideset.h"
+#include "InputParameters.h"
+#include "MooseTypes.h"
+#include "MooseMesh.h"
+
+#include "libmesh/mesh.h"
+#include "libmesh/remote_elem.h"
+
+#include <set>
+
+registerMooseObject("MooseApp", LowerDBlockFromSideset);
+
+template <>
+InputParameters
+validParams<LowerDBlockFromSideset>()
+{
+  InputParameters params = validParams<MeshModifier>();
+  params.addRequiredParam<SubdomainID>("new_block_id", "The lower dimensional block id to create");
+  params.addParam<SubdomainName>("new_block_name",
+                                 "The lower dimensional block name to create (optional)");
+  params.addRequiredParam<std::vector<BoundaryID>>(
+      "sidesets", "The sidesets from which to create the new block");
+  return params;
+}
+
+LowerDBlockFromSideset::LowerDBlockFromSideset(const InputParameters & parameters)
+  : MeshModifier(parameters),
+    _new_block_id(getParam<SubdomainID>("new_block_id")),
+    _sidesets(getParam<std::vector<BoundaryID>>("sidesets"))
+{
+}
+
+// Used to temporarily store information about which lower-dimensional
+// sides to add and what subdomain id to use for the added sides.
+struct ElemSideDouble
+{
+  ElemSideDouble(Elem * elem_in, unsigned short int side_in) : elem(elem_in), side(side_in) {}
+
+  Elem * elem;
+  unsigned short int side;
+};
+
+void
+LowerDBlockFromSideset::modify()
+{
+  // Reference the the libMesh::MeshBase
+  MeshBase & mesh = _mesh_ptr->getMesh();
+
+  // A reference to the Mesh's BoundaryInfo object
+  BoundaryInfo & bi = mesh.get_boundary_info();
+
+  // Data structures to store results of calling BoundaryInfo::build_side_list().
+  std::vector<dof_id_type> element_id_list;
+  std::vector<unsigned short int> side_list;
+  std::vector<boundary_id_type> bc_id_list;
+
+  // Get lists of (elem, side, id) triplets.
+  bi.build_side_list(element_id_list, side_list, bc_id_list);
+
+  std::set<BoundaryID> sidesets(_sidesets.begin(), _sidesets.end());
+
+  std::vector<ElemSideDouble> element_sides_on_boundary;
+  for (unsigned int i = 0; i < element_id_list.size(); ++i)
+    if (sidesets.count(bc_id_list[i]))
+      element_sides_on_boundary.push_back(
+          ElemSideDouble(mesh.elem_ptr(element_id_list[i]), side_list[i]));
+
+  for (auto && element_side : element_sides_on_boundary)
```

Comment:
I don't think we typically use rvalue references in our range-based loops, either `const auto &` or `auto &` is the usual convention.